In [1]:
#import pandas and numpy
import pandas as pd
import numpy as np

# Import data with Pandas

In [6]:
data = ("C:/datasets/BlackFriday.csv")
df = pd.read_csv(data)

# Exploratory Data Analysis
The next step is to conduct exploratory data analysis.


In [ ]:
#Check the type of dataframe
type(df)

In [13]:
#Check shape of dataframe
df.shape

(550068, 12)

There are 537577 rows and 12 columns in the dataset.


In [14]:
#View the first five rows of the dataframe
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [15]:
#View concise summary of dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


# Handle missing values with pandas

In [16]:
df.isnull().sum()

User_ID                            0
Product_ID                         0
Gender                             0
Age                                0
Occupation                         0
City_Category                      0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
dtype: int64

In [17]:
df.isna().sum()

User_ID                            0
Product_ID                         0
Gender                             0
Age                                0
Occupation                         0
City_Category                      0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
dtype: int64

# Encode missing numerical values

# Handle missing numerical values


In [23]:
df = df.ffill()


In [24]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            1
Product_Category_3            1
Purchase                      0
dtype: int64

In [25]:
df[["Product_Category_2","Product_Category_3"]].head()

Product_Category_2  Product_Category_3
0                 NaN                 NaN
1                 6.0                14.0
2                 6.0                14.0
3                14.0                14.0
4                14.0                14.0

In [27]:
df = df.bfill()

In [28]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
dtype: int64

# Check with ASSERT statement

In [30]:
assert pd.notnull(df).all().all()

The above command does not throw any AssertionError. So, it is confirmed that there are no missing values in the dataframe.

# Indexing and slicing in pandas

In [31]:
#Make a copy of dataframe
df1=df.copy()

In [32]:
#Select first row of dataframe using loc
df1.loc[0]

User_ID                         1000001
Product_ID                    P00069042
Gender                                F
Age                                0-17
Occupation                           10
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        0
Product_Category_1                    3
Product_Category_2                  6.0
Product_Category_3                 14.0
Purchase                           8370
Name: 0, dtype: object

In [33]:
#select first five rows for a specific column
df1.loc[:,'Purchase'].head()

0     8370
1    15200
2     1422
3     1057
4     7969
Name: Purchase, dtype: int64

In [34]:
#select first row of dataframe using iloc
df1.iloc[0]

User_ID                         1000001
Product_ID                    P00069042
Gender                                F
Age                                0-17
Occupation                           10
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        0
Product_Category_1                    3
Product_Category_2                  6.0
Product_Category_3                 14.0
Purchase                           8370
Name: 0, dtype: object

In [35]:
#Select last row of dataframe
df1.iloc[-1]

User_ID                         1006039
Product_ID                    P00371644
Gender                                F
Age                               46-50
Occupation                            0
City_Category                         B
Stay_In_Current_City_Years           4+
Marital_Status                        1
Product_Category_1                   20
Product_Category_2                  2.0
Product_Category_3                 11.0
Purchase                            490
Name: 550067, dtype: object

# Indexing first occurrence of maximum or minimum values with idxmax() and idxmin()

In [36]:
# Get index of first occurence of maximum Purchase value 
df1['Purchase'].idxmax()

87440

In [37]:
# Get the row with the maximum Purchase value 
df1.loc[df1['Purchase'].idxmax()]

User_ID                         1001474
Product_ID                    P00052842
Gender                                M
Age                               26-35
Occupation                            4
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        1
Product_Category_1                   10
Product_Category_2                 15.0
Product_Category_3                  8.0
Purchase                          23961
Name: 87440, dtype: object

# Indexing a single value with at() and iat()

In [39]:
# Get value at 1st row and Purchase column pair
df1.at[1, 'Purchase']

15200

In [40]:
# Get value at 1st row and 11th column pair

df1.iat[1, 11]

15200

# Indexing and reindexing in pandas

# Create a new dataframe

In [24]:
food = pd.DataFrame({'Place':['Home', 'Home', 'Hotel', 'Hotel'],
                   'Time': ['Lunch', 'Dinner', 'Lunch', 'Dinner'],
                   'Food':['Soup', 'Rice', 'Soup', 'Chapati'],
                   'Price($)':[10, 20, 30, 40]})
food

Place    Time     Food  Price($)
0   Home   Lunch     Soup        10
1   Home  Dinner     Rice        20
2  Hotel   Lunch     Soup        30
3  Hotel  Dinner  Chapati        40

In [27]:
# set_index() method which takes a column place as index
food_index1=food.set_index("Place")
food_index1

Time     Food  Price($)
Place                           
Home    Lunch     Soup        10
Home   Dinner     Rice        20
Hotel   Lunch     Soup        30
Hotel  Dinner  Chapati        40

In [28]:
food_index2=food.set_index("Place")
food_index2

Time     Food  Price($)
Place                           
Home    Lunch     Soup        10
Home   Dinner     Rice        20
Hotel   Lunch     Soup        30
Hotel  Dinner  Chapati        40

In [29]:
#Set multiple indexes
food_index2=food.set_index(["Place","Time"])
food_index2

Food  Price($)
Place Time                     
Home  Lunch      Soup        10
      Dinner     Rice        20
Hotel Lunch      Soup        30
      Dinner  Chapati        40

In [30]:
# Reset the index- Sets a simple integer index. This is the inverse operation of set_index()
food_index2.reset_index()

Place    Time     Food  Price($)
0   Home   Lunch     Soup        10
1   Home  Dinner     Rice        20
2  Hotel   Lunch     Soup        30
3  Hotel  Dinner  Chapati        40

# Hierarchical indexing or MultiIndex

The MultiIndex object is the hierarchical analogue of the standard index object which stores the axis labels in pandas objects. A MultiIndex is an array of tuples where each tuple is unique. A MultiIndex can be created from a list of arrays (using MultiIndex.from_arrays()), an array of tuples (using MultiIndex.from_tuples()), a crossed set of iterables (using MultiIndex.from_product()), or a DataFrame (using MultiIndex.from_frame()). The Index constructor will attempt to return a MultiIndex when it is passed a list of tuples.


In [32]:
#To demonstrate, lets create a df
sales=pd.DataFrame([['books','online', 200, 50],['books','retail', 250, 75], 
                    ['toys','online', 100, 20],['toys','retail', 140, 30],
                    ['watches','online', 500, 100],['watches','retail', 600, 150],
                    ['computers','online', 1000, 200],['computers','retail', 1200, 300],
                    ['laptops','online', 1100, 400],['laptops','retail', 1400, 500],
                    ['smartphones','online', 600, 200],['smartphones','retail', 800, 250]],
                    columns=['Items', 'Mode', 'Price', 'Profit'])
sales



Items    Mode  Price  Profit
0         books  online    200      50
1         books  retail    250      75
2          toys  online    100      20
3          toys  retail    140      30
4       watches  online    500     100
5       watches  retail    600     150
6     computers  online   1000     200
7     computers  retail   1200     300
8       laptops  online   1100     400
9       laptops  retail   1400     500
10  smartphones  online    600     200
11  smartphones  retail    800     250

# Create the hierarchical index in pandas

We can create a hierarchical index in pandas using the set_index() function which is used for indexing. First the data is indexed on Items and then on Mode column as follows:

In [36]:
sales1=sales.set_index(["Items","Mode"])
sales1

Price  Profit
Items       Mode                 
books       online    200      50
            retail    250      75
toys        online    100      20
            retail    140      30
watches     online    500     100
            retail    600     150
computers   online   1000     200
            retail   1200     300
laptops     online   1100     400
            retail   1400     500
smartphones online    600     200
            retail    800     250

# View index in hierarchical index

In [38]:
sales1.index

MultiIndex([(      'books', 'online'),
            (      'books', 'retail'),
            (       'toys', 'online'),
            (       'toys', 'retail'),
            (    'watches', 'online'),
            (    'watches', 'retail'),
            (  'computers', 'online'),
            (  'computers', 'retail'),
            (    'laptops', 'online'),
            (    'laptops', 'retail'),
            ('smartphones', 'online'),
            ('smartphones', 'retail')],
           names=['Items', 'Mode'])

# Swap the column in hierarchical index

In [39]:
sales2=sales1.swaplevel("Mode","Items")
sales2

Price  Profit
Mode   Items                     
online books          200      50
retail books          250      75
online toys           100      20
retail toys           140      30
online watches        500     100
retail watches        600     150
online computers     1000     200
retail computers     1200     300
online laptops       1100     400
retail laptops       1400     500
online smartphones    600     200
retail smartphones    800     250

# Sorting in pandas

# 1. Sorting by label


 We can use the sort_index() method to sort the object by labels. DataFrame can be sorted by passing the axis arguments and the order of sorting. By default, sorting is done on row labels in ascending order.

In [42]:
df2.sort_index()

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      NaN                 NaN      8370  
1                      6.0                14.0     15200  
2                      NaN                 NaN      1422  
3                     14.0                 NaN      1057  
4                      NaN                 NaN      7969  
...                    ...                 ...       ...  
550063                 NaN                 NaN       368  
550064                 NaN                 NaN       371  
550065                 NaN                 NaN       137  
550066                 NaN                 NaN       365  
550067                 NaN                 NaN       490  

[550068 rows x 12 columns]

In [44]:
#Sort in reverse order
df2.sort_index(ascending=False)

User_ID Product_ID Gender    Age  Occupation City_Category  \
550067  1006039  P00371644      F  46-50           0             B   
550066  1006038  P00375436      F    55+           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550064  1006035  P00375436      F  26-35           1             C   
550063  1006033  P00372445      M  51-55          13             B   
...         ...        ...    ...    ...         ...           ...   
4       1000002  P00285442      M    55+          16             C   
3       1000001  P00085442      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
0       1000001  P00069042      F   0-17          10             A   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550067                         4+               1                  20   
550066                          2               0                  20   
550065                         4+               1                  20   
550064                          3               0                  20   
550063                          1               1                  20   
...                           ...             ...                 ...   
4                              4+               0                   8   
3                               2               0                  12   
2                               2               0                  12   
1                               2               0                   1   
0                               2               0                   3   

        Product_Category_2  Product_Category_3  Purchase  
550067                 NaN                 NaN       490  
550066                 NaN                 NaN       365  
550065                 NaN                 NaN       137  
550064                 NaN                 NaN       371  
550063                 NaN                 NaN       368  
...                    ...                 ...       ...  
4                      NaN                 NaN      7969  
3                     14.0                 NaN      1057  
2                      NaN                 NaN      1422  
1                      6.0                14.0     15200  
0                      NaN                 NaN      8370  

[550068 rows x 12 columns]

# Sorting by columns

By passing the axis argument with a value 0 or 1, the sorting can be done on the row or column labels.
The default value of axis=0. In this case, sorting can be done by rows.
If we set axis=1, sorting is done by columns.

In [45]:
# Sort row wise
df.sort_index(axis=1)

Age City_Category Gender  Marital_Status  Occupation  \
0        0-17             A      F               0          10   
1        0-17             A      F               0          10   
2        0-17             A      F               0          10   
3        0-17             A      F               0          10   
4         55+             C      M               0          16   
...       ...           ...    ...             ...         ...   
550063  51-55             B      M               1          13   
550064  26-35             C      F               0           1   
550065  26-35             B      F               1          15   
550066    55+             C      F               0           1   
550067  46-50             B      F               1           0   

        Product_Category_1  Product_Category_2  Product_Category_3 Product_ID  \
0                        3                 NaN                 NaN  P00069042   
1                        1                 6.0                14.0  P00248942   
2                       12                 NaN                 NaN  P00087842   
3                       12                14.0                 NaN  P00085442   
4                        8                 NaN                 NaN  P00285442   
...                    ...                 ...                 ...        ...   
550063                  20                 NaN                 NaN  P00372445   
550064                  20                 NaN                 NaN  P00375436   
550065                  20                 NaN                 NaN  P00375436   
550066                  20                 NaN                 NaN  P00375436   
550067                  20                 NaN                 NaN  P00371644   

        Purchase Stay_In_Current_City_Years  User_ID  
0           8370                          2  1000001  
1          15200                          2  1000001  
2           1422                          2  1000001  
3           1057                          2  1000001  
4           7969                         4+  1000002  
...          ...                        ...      ...  
550063       368                          1  1006033  
550064       371                          3  1006035  
550065       137                         4+  1006036  
550066       365                          2  1006038  
550067       490                         4+  1006039  

[550068 rows x 12 columns]

In [46]:
# Sort column wise
df.sort_index(axis=0)

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      NaN                 NaN      8370  
1                      6.0                14.0     15200  
2                      NaN                 NaN      1422  
3                     14.0                 NaN      1057  
4                      NaN                 NaN      7969  
...                    ...                 ...       ...  
550063                 NaN                 NaN       368  
550064                 NaN                 NaN       371  
550065                 NaN                 NaN       137  
550066                 NaN                 NaN       365  
550067                 NaN                 NaN       490  

[550068 rows x 12 columns]

# 2. Sorting by values

The second method of sorting is sorting by values. Pandas provides sort_values() method to sort by values. It accepts a 'by' argument which will use the column name of the DataFrame with which the values are to be sorted.

In [48]:
df2.sort_values(by=["Product_Category_1"])

User_ID Product_ID Gender    Age  Occupation City_Category  \
271814  1005880  P00016042      M  26-35           1             A   
208659  1002109  P00298942      M  26-35          16             B   
436707  1001231  P00334242      M  26-35          12             C   
108508  1004685  P00025442      M  36-45           1             B   
208658  1002109  P00062842      M  26-35          16             B   
...         ...        ...    ...    ...         ...           ...   
547638  1002549  P00375436      M    55+          13             C   
547640  1002553  P00375436      M  26-35           7             C   
547642  1002556  P00371644      M  26-35           4             C   
547644  1002558  P00375436      M    55+          17             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
271814                          1               1                   1   
208659                          2               0                   1   
436707                          1               0                   1   
108508                          2               1                   1   
208658                          2               0                   1   
...                           ...             ...                 ...   
547638                          3               1                  20   
547640                          0               0                  20   
547642                          2               0                  20   
547644                          3               1                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
271814                16.0                 NaN      8322  
208659                 5.0                 NaN      7823  
436707                 8.0                 NaN     15803  
108508                 2.0                 9.0     15647  
208658                 2.0                13.0     11643  
...                    ...                 ...       ...  
547638                 NaN                 NaN       613  
547640                 NaN                 NaN       240  
547642                 NaN                 NaN       120  
547644                 NaN                 NaN       373  
550067                 NaN                 NaN       490  

[550068 rows x 12 columns]

# Sort by multiple columns

In [49]:

df2.sort_values(by=['Product_Category_1', 'Product_Category_2'])

User_ID Product_ID Gender    Age  Occupation City_Category  \
5       1000003  P00193542      M  26-35          15             A   
13      1000005  P00145042      M  26-35          20             A   
39      1000010  P00221342      F  36-45           1             B   
48      1000011  P00110842      F  26-35           1             C   
64      1000015  P00042142      M  26-35           7             A   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
5                               3               0                   1   
13                              1               1                   1   
39                             4+               1                   1   
48                              1               0                   1   
64                              1               0                   1   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
5                      2.0                 NaN     15227  
13                     2.0                 5.0     15665  
39                     2.0                 5.0     15705  
48                     2.0                 5.0     19327  
64                     2.0                 6.0     11458  
...                    ...                 ...       ...  
550063                 NaN                 NaN       368  
550064                 NaN                 NaN       371  
550065                 NaN                 NaN       137  
550066                 NaN                 NaN       365  
550067                 NaN                 NaN       490  

[550068 rows x 12 columns]

# Sort in descending order

In [50]:

df2.sort_values(by='Product_Category_1', ascending=False)

User_ID Product_ID Gender    Age  Occupation City_Category  \
550067  1006039  P00371644      F  46-50           0             B   
547652  1002572  P00375436      M  36-45          14             C   
547670  1002593  P00375436      M    55+          16             C   
547668  1002590  P00371644      M  18-25           4             A   
547667  1002589  P00375436      M  26-35           0             C   
...         ...        ...    ...    ...         ...           ...   
35645   1005504  P00182342      M  46-50           7             B   
255039  1003384  P00111142      M  36-45           7             C   
438811  1001557  P00063342      F  18-25           4             C   
438812  1001558  P00110942      M  18-25           6             A   
56674   1002761  P00129342      M  36-45           6             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550067                         4+               1                  20   
547652                          1               0                  20   
547670                          2               0                  20   
547668                          0               0                  20   
547667                          3               0                  20   
...                           ...             ...                 ...   
35645                           1               1                   1   
255039                          1               0                   1   
438811                          1               0                   1   
438812                          1               1                   1   
56674                           0               0                   1   

        Product_Category_2  Product_Category_3  Purchase  
550067                 NaN                 NaN       490  
547652                 NaN                 NaN       474  
547670                 NaN                 NaN       138  
547668                 NaN                 NaN       592  
547667                 NaN                 NaN       370  
...                    ...                 ...       ...  
35645                  5.0                 6.0     11691  
255039                15.0                16.0     19018  
438811                 2.0                 9.0     15789  
438812                 2.0                 NaN     19421  
56674                  5.0                15.0     15489  

[550068 rows x 12 columns]

# Categorical data in pandas

In [52]:
df3= df.copy()

In [53]:
df3.dtypes

User_ID                         int64
Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                        int64
dtype: object

We can see that our dataset has 5 categorical variables. They are Product_ID, Gender, Age, City_Category and Stay_In_Current_City_Years. They have data types as object.

# Description of categorical data

The describe() method on categorical data will produce similar output to a Series or DataFrame of type string.

In [54]:
df3['Gender'].describe()

count     550068
unique         2
top            M
freq      414259
Name: Gender, dtype: object

The Gender category has 550068 counts, 2 unique values and frequency of top value M is 414259.

In [55]:
df3['Age'].describe()

count     550068
unique         7
top        26-35
freq      219587
Name: Age, dtype: object

There are 7 unique categories in Age variable. The most frequent category is 26-35 with frequency count of 219587.

In [57]:
df3['City_Category'].describe()

count     550068
unique         3
top            B
freq      231173
Name: City_Category, dtype: object

There are 3 unique categories in City_Category variable. The most frequent category is B with frequency count of  231173.

# Unique values in categorical data

In [58]:
df3['Gender'].unique()

array(['F', 'M'], dtype=object)

In [59]:
df3['Age'].unique()

array(['0-17', '55+', '26-35', '46-50', '51-55', '36-45', '18-25'],
      dtype=object)

# Frequency counts of categorical data

Series.value_counts() will return the frequency counts of the categories present in the series.

In [60]:
df3['Gender'].value_counts()

Gender
M    414259
F    135809
Name: count, dtype: int64

In [61]:
df3['City_Category'].value_counts()

City_Category
B    231173
C    171175
A    147720
Name: count, dtype: int64

In [62]:
df3['Gender'].value_counts(ascending=True)

Gender
F    135809
M    414259
Name: count, dtype: int64

In [64]:
df3['City_Category'].value_counts(ascending=True)

City_Category
A    147720
C    171175
B    231173
Name: count, dtype: int64

# Basic functionality in pandas

# Series basic functionality

The following table lists the important attributes or methods in Series basic functionality.

axes - Returns a list of the row axis labels

dtype - Returns the dtype of the object.

empty - Returns True if series is empty.

ndim - Returns the number of dimensions of the underlying data, by definition 1.

size - Returns the number of elements in the underlying data.

values - Returns the Series as ndarray.

head() - Returns the first n rows.

tail() - Returns the last n rows.

# Dataframe basic functionality

The following tables lists the important attributes or methods in Dataframe basic functionality.

T - Transposes rows and columns.

axes - Returns a list with the row axis labels and column axis labels as the only members.

dtypes - Returns the dtypes in this object.

empty - True if NDFrame is entirely empty [no items]; if any of the axes are of length 0.

ndim - Number of axes / array dimensions.

shape - Returns a tuple representing the dimensionality of the Dataframe.

size - Number of elements in the NDFrame.

values - Numpy representation of NDFrame.

head() - Returns the first n rows.

tail() - Returns last n rows.

# Descriptive statistics in pandas

# Functions and descriptions

The following table list down the important functions under Descriptive Statistics in Python Pandas.

1 count() - Number of non-null observations

2 sum() - Sum of values

3 mean() - Mean of values

4 median() - Median of values

5 mode() - Mode of values

6 std() - Standard deviation of the values

7 min() - Minimum value

8 max() - Maximum value

9 abs() - Absolute value

10 prod() - Product of values

11 cumsum() - Cumulative sum

12 cumprod() - Cumulative product

The dataframe is a heterogeneous data structure. So, the different column values have different data types. Generic operations don't work with all functions.

Functions like sum(), cumsum() work with both numeric and character (or) string data elements without any error. In practice, character aggregations are never used generally. These functions do not throw any exception.

Functions like abs(), cumprod() throw exception when the dataframe contains character or string data because such operations cannot be performed.

In [75]:
df4=df.copy()

df4.max(0)

User_ID                        1006040
Product_ID                    P0099942
Gender                               M
Age                                55+
Occupation                          20
City_Category                        C
Stay_In_Current_City_Years          4+
Marital_Status                       1
Product_Category_1                  20
Product_Category_2                18.0
Product_Category_3                18.0
Purchase                         23961
dtype: object

# Summarizing data

The describe() function computes the summary statistics of the numerical columns in the dataframe.
This function gives the mean, std and IQR values. It excludes the character columns and gives summary about numeric columns. It includes the argument which is used to pass necessary information regarding what columns need to be considered for summarizing. It takes the list of values; by default, 'number'.

In [77]:

df4.describe()

User_ID     Occupation  Marital_Status  Product_Category_1  \
count  5.500680e+05  550068.000000   550068.000000       550068.000000   
mean   1.003029e+06       8.076707        0.409653            5.404270   
std    1.727592e+03       6.522660        0.491770            3.936211   
min    1.000001e+06       0.000000        0.000000            1.000000   
25%    1.001516e+06       2.000000        0.000000            1.000000   
50%    1.003077e+06       7.000000        0.000000            5.000000   
75%    1.004478e+06      14.000000        1.000000            8.000000   
max    1.006040e+06      20.000000        1.000000           20.000000   

       Product_Category_2  Product_Category_3       Purchase  
count       376430.000000       166821.000000  550068.000000  
mean             9.842329           12.668243    9263.968713  
std              5.086590            4.125338    5023.065394  
min              2.000000            3.000000      12.000000  
25%              5.000000            9.000000    5823.000000  
50%              9.000000           14.000000    8047.000000  
75%             15.000000           16.000000   12054.000000  
max             18.000000           18.000000   23961.000000

# Statistical functions in pandas

Statistical functions help us to understand and analyze the behavior of data. In this section, I will discuss few statistical functions, which we can apply on Pandas objects.

# Percent_change

Series, datFrames and panel, all have the function pct_change(). This function compares every element with its prior element and computes the change percentage.
By default, the pct_change() operates on columns; if you want to apply the same row wise, then use axis=1() argument.

# Covariance

Covariance is applied on series data. The series object has a method cov() to compute covariance between series objects. NA values will be excluded automatically.
Series.cov() can be used to compute covariance between series (excluding missing values).
Analogously, dataFrame.cov() to compute pairwise covariances among the series in the dataFrame, also excluding NA/null values.

In [80]:
df5=df.copy()

#For numeric columns, set numeric_only to True
df5.cov(numeric_only=True)



User_ID  Occupation  Marital_Status  \
User_ID             2.984573e+06 -270.113921       17.367619   
Occupation         -2.701139e+02   42.545100        0.077882   
Marital_Status      1.736762e+01    0.077882        0.241838   
Product_Category_1  2.600801e+01   -0.195578        0.038497   
Product_Category_2  1.346196e+01   -0.012700        0.037819   
Product_Category_3  2.443929e+01    0.354959        0.039400   
Purchase            4.092159e+04  682.554656       -1.144629   

                    Product_Category_1  Product_Category_2  \
User_ID                      26.008008           13.461964   
Occupation                   -0.195578           -0.012700   
Marital_Status                0.038497            0.037819   
Product_Category_1           15.493760            9.044359   
Product_Category_2            9.044359           25.873394   
Product_Category_3            2.438836           10.092953   
Purchase                  -6795.650007        -5538.125114   

                    Product_Category_3      Purchase  
User_ID                      24.439292  4.092159e+04  
Occupation                    0.354959  6.825547e+02  
Marital_Status                0.039400 -1.144629e+00  
Product_Category_1            2.438836 -6.795650e+03  
Product_Category_2           10.092953 -5.538125e+03  
Product_Category_3           17.018411 -4.613754e+02  
Purchase                   -461.375426  2.523119e+07

# Correlation

Correlation shows the linear relationship between any two array of values (series). There are multiple methods to compute the correlation. These methods are listed below:-
Method name Description
pearson (default) - Standard correlation coefficient
kendall - Kendall Tau correlation coefficient
spearman - Spearman rank correlation coefficient
All of these are currently computed using pairwise complete observations.
Any non-numeric columns will be automatically excluded from the correlation calculation.

In [82]:
# view the correlation

df5.corr(numeric_only=True)

User_ID  Occupation  Marital_Status  Product_Category_1  \
User_ID             1.000000   -0.023971        0.020443            0.003825   
Occupation         -0.023971    1.000000        0.024280           -0.007618   
Marital_Status      0.020443    0.024280        1.000000            0.019888   
Product_Category_1  0.003825   -0.007618        0.019888            1.000000   
Product_Category_2  0.001529   -0.000384        0.015138            0.540583   
Product_Category_3  0.003419    0.013263        0.019473            0.229678   
Purchase            0.004716    0.020833       -0.000463           -0.343703   

                    Product_Category_2  Product_Category_3  Purchase  
User_ID                       0.001529            0.003419  0.004716  
Occupation                   -0.000384            0.013263  0.020833  
Marital_Status                0.015138            0.019473 -0.000463  
Product_Category_1            0.540583            0.229678 -0.343703  
Product_Category_2            1.000000            0.543649 -0.209918  
Product_Category_3            0.543649            1.000000 -0.022006  
Purchase                     -0.209918           -0.022006  1.000000

# Data Ranking

Data Ranking produces ranking for each element in the array of elements. In case of ties, assigns the mean rank.
The rank() method produces a data ranking with ties being assigned the mean of the ranks (by default) for the group.
The rank() is also a dataframe method and can rank either the rows (axis=0) or the columns (axis=1). NaN values are excluded from the ranking.
It optionally takes a parameter ascending which true by default. If it is set to false, data is ranked in descending order, with larger values assigned a smaller rank.
The rank() supports different tie-breaking methods, specified with the method parameter as follows:-

average - average rank of tied group

min - lowest rank in the group

max - highest rank in the group

first - ranks assigned in the order they appear in the array

In [84]:
# view the top 25 rows of ranked dataframe
df5_numeric = df5.select_dtypes(include='number')
df5_numeric.rank(axis=1).head(25)


User_ID  Occupation  Marital_Status  Product_Category_1  \
0       5.0         3.0             1.0                 2.0   
1       7.0         4.0             1.0                 2.0   
2       5.0         2.0             1.0                 3.0   
3       6.0         2.0             1.0                 3.0   
4       5.0         3.0             1.0                 2.0   
5       6.0         4.0             1.0                 2.0   
6       7.0         3.0             1.5                 1.5   
7       6.0         3.0             1.5                 1.5   
8       6.0         3.0             1.5                 1.5   
9       5.0         3.0             1.0                 2.0   
10      6.0         4.0             1.0                 2.0   
11      5.0         3.0             1.0                 2.0   
12      5.0         3.0             1.0                 2.0   
13      7.0         5.0             1.5                 1.5   
14      7.0         4.0             1.0                 2.0   
15      6.0         4.0             1.0                 2.0   
16      7.0         5.0             1.0                 2.0   
17      6.0         3.0             1.0                 2.0   
18      7.0         2.0             2.0                 2.0   
19      7.0         4.0             1.5                 1.5   
20      6.0         3.0             1.0                 2.0   
21      5.0         3.0             1.0                 2.0   
22      5.0         3.0             1.0                 2.0   
23      5.0         3.0             1.0                 2.0   
24      7.0         4.0             1.5                 1.5   

    Product_Category_2  Product_Category_3  Purchase  
0                  NaN                 NaN       4.0  
1                  3.0                 5.0       6.0  
2                  NaN                 NaN       4.0  
3                  4.0                 NaN       5.0  
4                  NaN                 NaN       4.0  
5                  3.0                 NaN       5.0  
6                  4.0                 5.0       6.0  
7                  4.0                 NaN       5.0  
8                  4.0                 NaN       5.0  
9                  NaN                 NaN       4.0  
10                 3.0                 NaN       5.0  
11                 NaN                 NaN       4.0  
12                 NaN                 NaN       4.0  
13                 3.0                 4.0       6.0  
14                 3.0                 5.0       6.0  
15                 3.0                 NaN       5.0  
16                 3.0                 4.0       6.0  
17                 4.0                 NaN       5.0  
18                 4.0                 5.0       6.0  
19                 3.0                 5.0       6.0  
20                 4.0                 NaN       5.0  
21                 NaN                 NaN       4.0  
22                 NaN                 NaN       4.0  
23                 NaN                 NaN       4.0  
24                 3.0                 5.0       6.0

# Common statistical functions


There are a number of common statistical functions. These are listed below:-

Method - Description

count() - Number of non-null observations

sum() - Sum of values

mean() - Mean of values

median() - Arithmetic median of values

min() - Minimum

max() - Maximum

std() - Standard deviation

var() - Variance

skew() - Skewness

kurt() - Kurtosis

quantile() - Quantile

apply() - Generic apply

cov() - Covariance

corr() - Correlation

The apply() function takes an extra func argument and performs generic rolling computations. The func argument should be a single function that produces a single value from an ndarray input.

# Aggregations in pandas

In [3]:
#import pandas and numpy
import pandas as pd
import numpy as np
data = ("C:/datasets/BlackFriday.csv")
df = pd.read_csv(data)
df6=df.copy



Apply aggregation on a single column of a dataframe

In [11]:
df6 = df.groupby('Purchase')
df6['Purchase'].aggregate(np.sum)



C:\Users\user\AppData\Local\Temp\ipykernel_6700\590232486.py:2: FutureWarning: The provided callable <function sum at 0x000001BFC06CB060> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  df6['Purchase'].aggregate(np.sum)


Purchase
12        1212
13        1378
14        1330
24        2832
25        2825
         ...  
23956    23956
23958    95832
23959    47918
23960    95840
23961    71883
Name: Purchase, Length: 18105, dtype: int64

Apply multiple functions on a single column of a dataframe

In [8]:
df6['Purchase'].aggregate([np.sum, np.mean])

C:\Users\user\AppData\Local\Temp\ipykernel_1688\2907118823.py:1: FutureWarning: The provided callable <function sum at 0x000002AE9DE060C0> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  df6['Purchase'].aggregate([np.sum, np.mean])
C:\Users\user\AppData\Local\Temp\ipykernel_1688\2907118823.py:1: FutureWarning: The provided callable <function mean at 0x000002AE9DE071A0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df6['Purchase'].aggregate([np.sum, np.mean])


sum     mean
Purchase                
12         1212     12.0
13         1378     13.0
14         1330     14.0
24         2832     24.0
25         2825     25.0
...         ...      ...
23956     23956  23956.0
23958     95832  23958.0
23959     47918  23959.0
23960     95840  23960.0
23961     71883  23961.0

[18105 rows x 2 columns]

Apply aggregation on multiple columns of a dataframe

# Pandas GroupBy operations

A groupby operation involves one of the following operations on the original object. They are as follows :

Splitting the Object

Applying a function

Combining the results

# Split Data into Groups

Pandas object can be split into any of their objects. There are multiple ways to split an object as follows :-

obj.groupby('key')

obj.groupby(['key1','key2'])

obj.groupby(key,axis=1)

In [9]:
df8=df.copy()

df8.groupby('Gender')

In [ ]:
# view groups of Gender column

df8.groupby('Gender').groups
df8.groupby('Gender').sum()

In [ ]:
#import pandas and numpy
import pandas as pd
import numpy as np
data = ("C:/datasets/BlackFriday.csv")
df = pd.read_csv(data)
df8=df.copy()
df8.groupby('Gender').sum()

# Transformations

Transformation on a group or a column returns an object that is indexed the same size of that is being grouped. Thus, the transform should return a result that is the same size as that of a group chunk.

In [1]:

df9=df.copy()
score = lambda x: (x - x.mean()) / x.std()*10
print(df9.groupby('Gender')['Purchase'].transform(score).head(5))

0    -0.764732
1    13.562236
2   -15.339224
3   -16.104867
4    -2.883881
Name: Purchase, dtype: float64


# Filtration

Filtration filters the data on a defined criteria and returns the subset of data. The filter() function is used to filter the data.

In [9]:
import pandas as pd
import numpy as np
data = "C:/datasets/BlackFriday.csv"
df = pd.read_csv(data)
df10 = df.copy()  
filtered_df = df10.groupby("Gender").filter(lambda x: len(x) > 4)
print(filtered_df.head())


   User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN   